test/batch_notifier_test.py

```diff
@@ -0,0 +1,562 @@
+            'Luigi: 4 failures in the last 60 minutes',
+            '- Task (3 failures)\n'
+            '- OtherTask (1 failure)'
+        )
+
+    def test_group_on_unbatched_params(self):
+        bn = BatchNotifier(batch_mode='unbatched_params')
+        bn.add_failure('Task(a=5, b=1)', 'Task', {'a': 5}, 'error')
+        bn.add_failure('Task(a=5, b=2)', 'Task', {'a': 5}, 'error')
+        bn.add_failure('Task(a=6, b=1)', 'Task', {'a': 6}, 'error')
+        bn.add_failure('Task(a=6, b=2)', 'Task', {'a': 6}, 'error')
+        bn.add_failure('Task(a=6, b=3)', 'Task', {'a': 6}, 'error')
+        bn.add_failure('Task(a=6, b=4)', 'Task', {'a': 6}, 'error')
+        bn.add_failure('OtherTask(a=5, b=1)', 'OtherTask', {'a': 5}, 'error')
+        bn.add_failure('OtherTask(a=6, b=1)', 'OtherTask', {'a': 6}, 'error')
+        bn.add_failure('OtherTask(a=6, b=2)', 'OtherTask', {'a': 6}, 'error')
+        bn.add_failure('OtherTask(a=6, b=3)', 'OtherTask', {'a': 6}, 'error')
+        bn.send_email()
+        self.check_email_send(
+            'Luigi: 10 failures in the last 60 minutes',
+            '- Task(a=6) (4 failures)\n'
+            '- OtherTask(a=6) (3 failures)\n'
+            '- Task(a=5) (2 failures)\n'
+            '- OtherTask(a=5) (1 failure)'
+        )
+
+    def test_include_one_expl(self):
+        bn = BatchNotifier(batch_mode='family', error_messages=1)
+        bn.add_failure('Task(a=1)', 'Task', {'a': 1}, 'error 1')
+        bn.add_failure('Task(a=2)', 'Task', {'a': 2}, 'error 2')
+        bn.add_failure('TaskB(a=1)', 'TaskB', {'a': 1}, 'error')
+
+        bn.send_email()
+        self.check_email_send(
+            'Luigi: 3 failures in the last 60 minutes',
+            '- Task (2 failures)\n'
+            '\n'
+            '      error 2\n'
+            '\n'
+            '- TaskB (1 failure)\n'
+            '\n'
+            '      error'
+        )
+
+    def test_include_two_expls(self):
+        bn = BatchNotifier(batch_mode='family', error_messages=2)
+        bn.add_failure('Task(a=1)', 'Task', {'a': 1}, 'error 1')
+        bn.add_failure('Task(a=2)', 'Task', {'a': 2}, 'error 2')
+        bn.add_failure('TaskB(a=1)', 'TaskB', {'a': 1}, 'error')
+
+        bn.send_email()
+        self.check_email_send(
+            'Luigi: 3 failures in the last 60 minutes',
+            '- Task (2 failures)\n'
+            '\n'
+            '      error 1\n'
+            '\n'
+            '      error 2\n'
+            '\n'
+            '- TaskB (1 failure)\n'
+            '\n'
+            '      error'
+        )
+
+    def test_limit_expl_length(self):
+        bn = BatchNotifier(batch_mode='family', error_messages=1, error_lines=2)
+        bn.add_failure('Task(a=1)', 'Task', {'a': '1'}, 'line 1\nline 2\nline 3\nline 4\n')
+        bn.send_email()
+        self.check_email_send(
+            'Luigi: 1 failure in the last 60 minutes',
+            '- Task (1 failure)\n'
+            '\n'
+            '      line 3\n'
```

Comment:
That's certainly a possibility. I hadn't thought about it too much. It might be better not to include that, as it could prevent grouping. I really find it helpful to group identical messages and have actually considered using fewer lines so more differences get cut out. If we include the number of stripped lines, I'll see something like

```
- TaskA (1 failure)

    ... 10 stripped lines ...
   mysql failure

- TaskB (1 failure)

    ... 11 stripped lines ...
    mysql failure

- TaskC

    ... 9 stripped lines ...
    mysql failure
```

instead of just

```
- TaskA (1 failure)
- TaskB (1 failure)
- TaskC (1 failure)

    mysql failure
```

because each task could have different trace lengths despite having the same underlying issue.